src/core/auth/wallet-authenticator.ts

```diff
@@ -0,0 +1,342 @@
+    }
+
+    // Check that the signing address is the claimed wallet address
+    const message = this.generateMessage(payload.payload);
+    const userAddress = this.recoverAddress(message, payload.signature);
+    if (userAddress.toLowerCase() !== payload.payload.address.toLowerCase()) {
+      throw new Error(
+        `Signer address ${userAddress.toLowerCase()} does not match payload address ${payload.payload.address.toLowerCase()}`,
+      );
+    }
+
+    return userAddress;
+  }
+
+  /**
+   * Generate Authentication Token
+   * @remarks Server-side function that generates a JWT token from the provided login request that the
+   * client-side wallet can use to authenticate to the server-side application.
+   *
+   * @param domain - The domain of the server-side application to authenticate to
+   * @param payload - The login payload to authenticate with
+   * @param options - Optional configuration options for the authentication request
+   * @returns A authentication token that can be used by the client to make authenticated requests
+   *
+   * @example
+   * ```javascript
+   * const domain = "thirdweb.com";
+   * const loginPayload = await sdk.auth.login(domain);
+   *
+   * // Generate a JWT token that can be sent to the client-side wallet and used for authentication
+   * const token = await sdk.auth.generate(domain, loginPayload);
+   * ```
+   */
+  public async generate(
+    domain: string,
+    payload: LoginPayload,
+    options?: AuthenticationOptions,
+  ): Promise<string> {
+    if (isBrowser()) {
+      throw new Error(
+        "Authentication tokens should not be generated in the browser.",
+      );
+    }
+
+    const parsedOptions = AuthenticationOptionsSchema.parse(options);
+
+    const userAddress = this.verify(domain, payload);
+    const adminAddress = await this.wallet.getAddress();
+    const payloadData = AuthenticationPayloadDataSchema.parse({
+      iss: adminAddress,
+      sub: userAddress,
+      aud: domain,
+      nbf: parsedOptions?.invalidBefore || new Date(),
+      exp:
+        parsedOptions?.expirationTime ||
+        new Date(Date.now() + 1000 * 60 * 60 * 5),
+      iat: new Date(),
+    });
+
+    const message = JSON.stringify(payloadData);
+    const signature = await this.wallet.sign(message);
+
+    // Header used for JWT token specifying hash algorithm
+    const header = {
+      // Specify ECDSA with SHA-256 for hashing algorithm
+      alg: "ES256",
+      typ: "JWT",
+    };
+
+    const encodedHeader = Buffer.from(JSON.stringify(header)).toString(
+      "base64",
+    );
+    const encodedData = Buffer.from(JSON.stringify(payloadData))
+      .toString("base64")
+      .replace(/=/g, "");
+    const encodedSignature = Buffer.from(signature).toString("base64");
+
+    // Generate a JWT token with base64 encoded header, payload, and signature
+    const token = `${encodedHeader}.${encodedData}.${encodedSignature}`;
+
+    return token;
+  }
+
+  /**
+   * Authenticate With Token
```

Comment:
wonder if we should specify 'auth token' here and below, since we use token for many other things in the sdk.

same for `generate` - think we can be more specific, something like `generateAuthToken`